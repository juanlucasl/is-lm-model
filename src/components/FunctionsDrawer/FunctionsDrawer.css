/* SPDX-License-Identifier: Apache-2.0 */
.functions-drawer {
  border-bottom: solid black 1px;
  border-left: solid black 1px;
  margin: 1rem 0;
}
