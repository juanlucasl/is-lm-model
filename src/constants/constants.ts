/* SPDX-License-Identifier: Apache-2.0 */

export const GRAPH_SIZE = 500;

export const MAX_VALUE = GRAPH_SIZE;
export const MIN_VALUE = -100;

/**
 * Propension Marginal a Consumir: Indica como una unidad más de renta disponible afecta al
 * consumo.
 */
export const PROPENSION_MARGINAL_A_CONSUMIR = 0.8;

export const SENSIBILIDAD_DE_LA_DEMANDA_MONETARIA_AL_INTERES = 0.5;

/**
 * Lo que consumirían los individuos si su renta disponible fuese igual a cero.
 */
export const CONSUMO_AUTONOMO = 100;
export const DEMANDA_BASE_DE_DINERO = CONSUMO_AUTONOMO;
