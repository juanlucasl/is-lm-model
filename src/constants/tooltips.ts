const CONSUMPTION_UP = "&uarr; Consumo (C)";
const CONSUMPTION_DOWN = "&darr; Consumo (C)";

const INTEREST_RATE_UP = "&uarr; Tipo de Interés (i)";
const INTEREST_RATE_DOWN = "&darr; Tipo de Interés (i)";

const OUTPUT_UP = "&uarr; Producción o PIB (Y) (En equilibrio Y = Z (Demanda de bienes))";
const OUTPUT_DOWN = "&darr; Producción o PIB (Y) (En equilibrio Y = Z (Demanda de bienes))";

const IS_RIGHT = "&rarr; La Curva IS se desplaza hacia la derecha, reflejando un aumento en la Demanda Agregada";
const IS_LEFT = "&larr; La Curva IS se desplaza hacia la izquierda, reflejando un descenso en la Demanda Agregada";

export const TAXES_INCREASE_DESCRIPTION = `
  <p>
    El Estado cobra más impuestos y la gente tiene menor excedente de renta para demandar
    bienes
    <ul>
      <li>${CONSUMPTION_DOWN}</li>
      <li>${INTEREST_RATE_DOWN}</li>
      <li>${OUTPUT_DOWN}</li>
      <li>${IS_LEFT}</li>
    </ul>
  </p>
`;

export const TAXES_DECREASE_DESCRIPTION = `
  <p>
    El Estado cobra menos impuestos y la gente tiene mayor excedente de renta para demandar
    bienes
    <ul>
      <li>${CONSUMPTION_UP}</li>
      <li>${INTEREST_RATE_UP}</li>
      <li>${OUTPUT_UP}</li>
      <li>${IS_RIGHT}</li>
    </ul>
  </p>
`;

export const INVESTMENT_INCREASE_DESCRIPTION = `
  <p>
    Mayor demanda de bienes de capital, por lo tanto, es mayor la Demanda Agregada
    <ul>
      <li>${CONSUMPTION_UP}</li>
      <li>${INTEREST_RATE_UP}</li>
      <li>${OUTPUT_UP}</li>
      <li>${IS_RIGHT}</li>
    </ul>
  </p>
`;

export const INVESTMENT_DECREASE_DESCRIPTION = `
  <p>
    Menor demanda de bienes de capital, por lo tanto, es menor la Demanda Agregada
    <ul>
      <li>${CONSUMPTION_DOWN}</li>
      <li>${INTEREST_RATE_DOWN}</li>
      <li>${OUTPUT_DOWN}</li>
      <li>${IS_LEFT}</li>
    </ul>
  </p>
`;

export const PUBLIC_EXPENDITURE_INCREASE_DESCRIPTION = `
  <p>
    El Estado demanda más bienes y servicios, por lo tanto, es mayor la Demanda Agregada
    <ul>
      <li>${CONSUMPTION_UP}</li>
      <li>${INTEREST_RATE_UP}</li>
      <li>${OUTPUT_UP}</li>
      <li>${IS_RIGHT}</li>
    </ul>
  </p>
`;

export const PUBLIC_EXPENDITURE_DECREASE_DESCRIPTION = `
  <p>
    El Estado demanda menos bienes y servicios, por lo tanto, es menor la Demanda Agregada
    <ul>
      <li>${CONSUMPTION_DOWN}</li>
      <li>${INTEREST_RATE_DOWN}</li>
      <li>${OUTPUT_DOWN}</li>
      <li>${IS_LEFT}</li>
    </ul>
  </p>
`;

export const MONEY_SUPPLY_INCREASE_DESCRIPTION = `
  <p>
    Mayor oferta de dinero hace bajar su precio, reflejado en el interés
    <ul>
      <li>${INTEREST_RATE_DOWN}</li>
    </ul>
  </p>
`;

export const MONEY_SUPPLY_DECREASE_DESCRIPTION = `
  <p>
    Menor oferta de dinero hace subir su precio, reflejado en el interés{" "}
    <ul>
      <li>${INTEREST_RATE_UP}</li>
    </ul>
  </p>
`;
