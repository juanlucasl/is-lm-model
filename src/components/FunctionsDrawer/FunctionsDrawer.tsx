/* SPDX-License-Identifier: Apache-2.0 */
import "./FunctionsDrawer.css";
import React, { useRef, useEffect } from "react";
import { GRAPH_SIZE } from "../../constants";
import { FunctionFormula } from "../../models";

interface FunctionsDrawerProps {
  functionsToDraw: FunctionFormula[];
  verticalAxisLabel?: string;
  horizontalAxisLabel?: string;
}

const colors = ["blue", "red", "green"];

/**
 * Receives the y coordinates of two functions at a horizontal point and returns true
 * if they intersect.
 *
 * The point is considered an intersection point between the functions if the difference
 * between y coordinates of each function at the point is lesser than 1.
 *
 * @param y1
 * @param y2
 */
const functionsDoIntersect = (y1: number, y2: number) => Math.abs(y1 - y2) < 1;

/**
 * Takes in an array of functionsToDraw and displays them on a canvas.
 *
 * @constructor
 */
const FunctionsDrawer: React.FC<FunctionsDrawerProps> = ({
  functionsToDraw,
  verticalAxisLabel = "Y",
  horizontalAxisLabel = "X",
}) => {
  const canvasRef = useRef<HTMLCanvasElement>(null);

  useEffect(() => {
    const canvas = canvasRef.current!;
    const ctx = canvas.getContext("2d")!;

    // Draw the functions
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    functionsToDraw.forEach((functionFormula, i) => {
      ctx.beginPath();
      ctx.moveTo(0, canvas.height - functionFormula(0));
      for (let j = 0; j <= canvas.height; j++) {
        ctx.lineTo(j, canvas.height - functionFormula(j));
      }

      // Set the color of the function using the index
      ctx.strokeStyle = colors[i % colors.length];

      ctx.stroke();
    });

    // Find the intersection points between functions
    for (let i = 0; i < functionsToDraw.length - 1; i++) {
      for (let j = i + 1; j < functionsToDraw.length; j++) {
        // Iterates through the horizontal axis of the canvas to check each horizontal pixel to
        // see if there is an intersection
        for (let horizontalPoint = 0; horizontalPoint <= canvas.width; horizontalPoint++) {
          // Calculate Y coordinate of each function at each point.
          const y1 = canvas.height - functionsToDraw[i](horizontalPoint);
          const y2 = canvas.height - functionsToDraw[j](horizontalPoint);
          if (functionsDoIntersect(y1, y2)) {
            // Set line style to be dotted and gray.
            ctx.strokeStyle = "gray";
            ctx.setLineDash([5, 5]);

            // Draw a line from the intersection point to the horizontal axis by moving the
            // draw path to the intersection (ctx.moveTo(horizontalPoint, y1)) and drawing a
            // line to the bottom of the canvas (ctx.lineTo(horizontalPoint, canvas.height))
            ctx.beginPath();
            ctx.moveTo(horizontalPoint, y1);
            ctx.lineTo(horizontalPoint, canvas.height);
            ctx.stroke();
            ctx.closePath();

            // Draw a line from the intersection point to the vertical axis by moving the path
            // to the intersection (ctx.moveTo(horizontalPoint, y1)) and drawing a line to the
            // left edge of the canvas (ctx.lineTo(0, y1))
            ctx.beginPath();
            ctx.moveTo(horizontalPoint, y1);
            ctx.lineTo(0, y1);
            ctx.stroke();
            ctx.closePath();

            // Reset style back to a solid line
            ctx.setLineDash([]);
          }
        }
      }
    }

    // Draw the axis labels
    ctx.font = "1.6rem sans-serif";
    ctx.fillText(verticalAxisLabel, 10, 15);
    ctx.fillText(horizontalAxisLabel, canvas.width - 15, canvas.height - 10);
  }, [functionsToDraw, horizontalAxisLabel, verticalAxisLabel]);

  return (
    <canvas
      aria-label="IS-LM"
      className="functions-drawer"
      height={GRAPH_SIZE}
      width={GRAPH_SIZE}
      ref={canvasRef}
      role="img"
    />
  );
};

export default FunctionsDrawer;
