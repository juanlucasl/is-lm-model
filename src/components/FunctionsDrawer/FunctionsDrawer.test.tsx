import React from "react";
import { render, screen } from "@testing-library/react";
import FunctionsDrawer from "./FunctionsDrawer";
import { GRAPH_SIZE } from "../../constants";

const mockFunctionsToDraw = [(x: number) => x, (x: number) => x * 2];

jest.mock("canvas", () => {
  const mCanvas = {
    getContext: jest.fn(),
  };
  return {
    createCanvas: jest.fn(() => mCanvas),
  };
});

describe("FunctionsDrawer", () => {
  it("renders correctly", () => {
    const { container } = render(<FunctionsDrawer functionsToDraw={mockFunctionsToDraw} />);
    expect(container).toMatchSnapshot();
  });

  it("renders the canvas element", () => {
    render(<FunctionsDrawer functionsToDraw={mockFunctionsToDraw} />);
    const canvas = screen.getByRole("img") as HTMLCanvasElement;
    expect(canvas).toBeInTheDocument();
    expect(canvas.width).toBe(GRAPH_SIZE);
    expect(canvas.height).toBe(GRAPH_SIZE);
  });
});
