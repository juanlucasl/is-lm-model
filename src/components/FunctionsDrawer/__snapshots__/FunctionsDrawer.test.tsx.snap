// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FunctionsDrawer renders correctly 1`] = `
<div>
  <canvas
    aria-label="IS-LM"
    class="functions-drawer"
    height="500"
    role="img"
    width="500"
  />
</div>
`;
